### ISSUE TITLE
Custom annotations on the created Nodes/Pods

### ISSUE THREAD

**Issue by dcherman** [zalando/zalenium#727]
I'm currently experimenting with running Zalenium in conjunction with [cluster-autoscaler](https://github.com/kubernetes/autoscaler) in order to get a cost effective but scalable grid up and running, but I've noticed that as nodes are added/removed from the cluster or as jobs complete, the autoscaler will attempt to move existing pods that may be executing tests to another node in order to allow it to scale down the cluster.  This is non-optimal since it will cause the test that's currently executing against that pod/browser to fail.

Since in my use case the pods are ephemeral and I want Zalenium to be solely responsible for creating/destroying them (with the exception of cordoning/draining a node, but that's separate), I'd like to add the `"cluster-autoscaler.kubernetes.io/safe-to-evict": "false"` annotation to the pods to disallow the cluster-autoscaler from managing these pods.

I'm not sure if an environment variable makes sense here like was used for the other config options since this one is inherently more complex since it's potentially not a singular value, but I'd like to get opinions on that.

Does something like this sound reasonable?

**Comment by diemol** [comment]
Hi @dcherman,

Is #662 what you mean?

**Comment by dcherman** [comment]
Hey @diemol,

Not exactly; so the node selector allows you to instruct k8s that a given pod may only be scheduled on a node containing that key/value pair, however the annotation that I'm referring to instructs the `cluster-autoscaler` that it may not evict a given pod for the purpose of scaling down the cluster since by default, it will attempt to eliminate resource waste.

[Here's some information](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node) about how to prevent the autoscaler from evicting/migrating specific pods, and [how the scale-down algoritm works](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-does-scale-down-work)

Does all of that make sense?

**Comment by pearj** [comment]
Sounds reasonable. It does sound like we need an easy way to customise how a pod gets created. Like if we could make a yaml template that is overridable with whatever feature you need to add to created pods. Or we can just allow a few injection points where you can add annotations, labels or node selectors.

**Comment by dcherman** [comment]
@pearj What about allowing users to do something like create an instance of a CustomResourceDefinition in order to provide you the config?

https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/

Essentially, you could allow users to define one or more resources like

```yaml
apiVersion: "zalenium.io/v1"
kind: ZaleniumBrowser
metadata:
  name: chrome-latest
  hub: zalenium
spec:
  pod:
    annotations:
      "cluster-autoscaler.kubernetes.io/safe-to-evict": "false"
    containers:
      - env:
        - name: SCREEN_HEIGHT
          value: "1080"
        - name: SCREEN_WIDTH
          value: "1920"
    image: elgalu/selenium
    imagePullPolicy: Always
    resources:
      limits:
        cpu: 1500m
      requests:
        cpu: "1"
        memory: 1Gi
  
```

You'd then be able to take those definitions and add on the additional settings/metadata that Zalenium needs to operate in order to get a pod template for creation.

This would also allow users a little more flexibility since you'd eventually be able to do something like supporting a min/max numbers of containers on a per resource basis, or allowing users to add custom annotations/capabilities to a given resource to allow certain ones to run with more/less resources depending on the needs of the application being tested.

**Comment by pearj** [comment]
Ahh that's clever, I've always wanted to use a CRD somehow.

**Comment by pearj** [comment]
kubernetes-client supports CRDs, which means this is definitely doable: https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/CRDExample.java

**Comment by diemol** [comment]
@dcherman would it be possible that you give us a hand with this? We don't have so much free time to tackle this.

**Comment by dcherman** [comment]
@diemol Sure, I'm willing to take a crack at this, sounds like a fun weekend project :)

**Comment by diemol** [comment]
I'm going to merge this into #654, and if someone has time to help us, we can reopen and continue the conversation here.

**Comment by amitbitcse** [comment]
I also encountered the same problem. While looking for an easier solution, I found that the selenium pods which are created by zalenium are not evicted by CA. Check below link.

https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node

CA might evict the zalenium pod and schedule it on another node during the node scale down process. This terminates all the existing running selenium pods. Adding "cluster-autoscaler.kubernetes.io/safe-to-evict": "false" annotation to the zalenium pod worked for me.
